rust/dyn-ddlog/src/ast/tree.rs

```diff
@@ -0,0 +1,651 @@
+use crate::ast::rewrites::{BlockSimplifier, ExpressionRewriter, NestingSimplifier};
+use std::{
+    fmt::{self, Debug, Write},
+    slice, vec,
+};
+
+pub use crate::resolve::Symbol;
+
+// TODO: Source locations
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Declaration<I> {
+    pub attributes: Vec<Attribute<I>>,
+    pub kind: DeclarationKind<I>,
+}
+
+impl<I> Declaration<I> {
+    pub const fn new(attributes: Vec<Attribute<I>>, kind: DeclarationKind<I>) -> Self {
+        Self { attributes, kind }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum DeclarationKind<I> {
+    Relation(Relation<I>),
+    Rule(Rule<I>),
+    Fact(Fact<I>),
+    Function(Function<I>),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Relation<I> {
+    pub kind: RelationKind,
+    pub name: I,
+    pub args: Vec<(I, Type<I>)>,
+}
+
+impl<I> Relation<I> {
+    pub const fn new(kind: RelationKind, name: I, args: Vec<(I, Type<I>)>) -> Self {
+        Self { kind, name, args }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum RelationKind {
+    Input,
+    Output,
+    Internal,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Rule<I> {
+    pub heads: Vec<RuleHead<I>>,
+    pub clauses: Vec<RuleClause<I>>,
+}
+
+impl<I> Rule<I> {
+    pub const fn new(heads: Vec<RuleHead<I>>, clauses: Vec<RuleClause<I>>) -> Self {
+        Self { heads, clauses }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct RuleHead<I> {
+    pub relation: I,
+    pub fields: Vec<Expr<I>>,
+}
+
+impl<I> RuleHead<I> {
+    pub const fn new(relation: I, fields: Vec<Expr<I>>) -> Self {
+        Self { relation, fields }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum RuleClause<I> {
+    Relation {
+        binding: Option<I>,
+        relation: I,
```

Comment:
I don't think so unless you can write this, function declaration's use of `Path` also confused me
```prolog
relation foo::Bar()
```

Reply:
No, but you can write `Rel(x) :- foo::Bar(x).` which is more relevant, as this is the body of a rule, not relation declaration.

Reply:
About this:

> ```prolog
> relation foo::Bar()
> ```

While this is not a valid input syntax, if you plan to flatten the module hierarchy as part of the compilation process, it may become valid intermediate syntax.